server/service/src/vaccination/insert/mod.rs

```diff
@@ -0,0 +1,331 @@
+use crate::{activity_log::activity_log_entry, service_provider::ServiceContext};
+
+use chrono::NaiveDate;
+use repository::{ActivityLogType, RepositoryError, Vaccination, VaccinationRowRepository};
+
+mod generate;
+mod validate;
+
+use generate::{generate, GenerateInput};
+use validate::validate;
+
+use super::query::get_vaccination;
+
+#[derive(PartialEq, Debug)]
+pub enum InsertVaccinationError {
+    VaccinationAlreadyExists,
+    EncounterDoesNotExist,
+    ProgramEnrolmentDoesNotExist,
+    VaccineCourseDoseDoesNotExist,
+    VaccinationAlreadyExistsForDose,
+    ClinicianDoesNotExist,
+    ReasonNotProvided,
+    StockLineNotProvided,
+    StockLineDoesNotExist,
+    ItemDoesNotBelongToVaccineCourse,
+    CreatedRecordNotFound,
+    DatabaseError(RepositoryError),
+}
+
+#[derive(PartialEq, Debug, Clone, Default)]
+pub struct InsertVaccination {
+    pub id: String,
+    pub encounter_id: String,
+    pub vaccine_course_dose_id: String,
+    pub vaccination_date: Option<NaiveDate>,
+    pub clinician_id: Option<String>,
+    pub comment: Option<String>,
+    pub given: bool,
+    pub stock_line_id: Option<String>,
+    pub not_given_reason: Option<String>,
+}
+
+pub fn insert_vaccination(
+    ctx: &ServiceContext,
+    store_id: &str,
+    input: InsertVaccination,
```

Comment:
Similar to my comment on last PR, do we want to add some logic here to ensure the "created_date" can't be before the minimum age for the course? (Would probably justify an additional Error value - `VaccintationTooEarly`?)

Can do in another PR -- maybe make an issue if you think worth it?

Reply:
Per other PR, product decision is to leave this open for clinician to deliver the dose before the exact age if need be 🙏